assignment2/assignment2.cpp

```diff
@@ -0,0 +1,119 @@
+#include <iostream>
+#include <vector>
+
+
+template <typename T>
+class BinaryTree
+{
+public:
+	struct Node{
+		T data;
+		Node *left_child;
+		Node *right_child;
+	};
+private:
+	Node *root;
+	
+	void remove_subtree(Node * node){
+		if(node == NULL){ return; }
+		Node *left= node->left_child;
+		Node *right = node->right_child;
+		delete node;
+		remove_subtree(left);
+		remove_subtree(right);
+	}
+
+	bool insert(const T &key, Node *tree, const T &new_data, bool right){
+		if(tree != NULL){
+
+			if(tree->data == key){
+				if(right){
+					tree->right_child = new Node;
+					tree->right_child->data = new_data;
+				} else {
+					tree->left_child = new Node;
+					tree->left_child->data = new_data;
+				}
+				return true;
+			}
+			if(insert(key, tree->left_child, new_data, right) || 
+				insert( key, tree->right_child, new_data, right)){
+				return true;
+			}
+		}
+		return false;
+	}
+
+	// Answer to Q1 from assignment2
+	bool print_ancestors(const T &key, Node *tree, std::vector<T> &ancestors){
+		if(tree != NULL){
+
+			if(tree->data == key){
+				return true;
+			}
+			if(print_ancestors(key, tree->left_child, ancestors) || 
+				print_ancestors( key, tree->right_child, ancestors)){
+				ancestors.push_back(tree->data);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	// Answer to Q2 from assignment2
+	Node* common_ancestor(const T& val1, const T&val2,Node* tree, bool& found){
```

Comment:
Also, it's useful to add to the comment (or better yet, the code :) ) the assumptions you are making, even if they are part of the problem statement, here for example that all keys are unique, and val1 != val2